src/t8_schemes/t8_default/t8_dtri_bits.c

```diff
@@ -1357,6 +1356,71 @@ t8_dtri_linear_id (const t8_dtri_t * t, int level)
   return id;
 }
 
+t8_linearidx_t
+t8_dtri_linear_id_with_level (const t8_dtri_t * t, int level, int stop)
+{
+  t8_linearidx_t      id = 0;
+  int8_t              type_temp = 0;
+  t8_dtri_cube_id_t   cid;
+  int                 i;
+  int                 exponent;
+  int                 my_level;
+
+  T8_ASSERT (0 <= level && level <= T8_DTRI_MAXLEVEL && level >= stop);
+  my_level = t->level;
+  exponent = 0;
+  /* If the given level is bigger than t's level
+   * we first fill up with the ids of t's descendants at t's
+   * origin with the same type as t */
+  if (level > my_level) {
+    exponent = (level - my_level) * T8_DTRI_DIM;
+  }
+  level = my_level;
+  type_temp = compute_type (t, level);
+  for (i = level; i > stop; i--) {
+    cid = compute_cubeid (t, i);
+    id |=
+      ((t8_linearidx_t) t8_dtri_type_cid_to_Iloc[type_temp][cid]) << exponent;
+    exponent += T8_DTRI_DIM;    /* multiply with 4 (2d) resp. 8  (3d) */
+    type_temp = t8_dtri_cid_type_to_parenttype[cid][type_temp];
+  }
+  return id;
+}
+
+void
+t8_dtri_init_linear_id_with_level (t8_dtri_t * t, t8_linearidx_t id,
+                                   const int start_level, const int end_level,
+                                   t8_dtri_type_t parenttype)
+{
+  int                 i;
+  int                 offset_coords, offset_index;
+  const int           children_m1 = T8_DTRI_CHILDREN - 1;
+  t8_linearidx_t      local_index;
+  t8_dtri_cube_id_t   cid;
+  t8_dtri_type_t      type;
+  T8_ASSERT (0 <= id
+             && id <= ((t8_linearidx_t) 1) << (T8_DTRI_DIM * end_level));
+
+  t->level = end_level;
+
+  type = parenttype;            /* This is the type of the parent triangle */
+  for (i = start_level; i <= end_level; i++) {
+    offset_coords = T8_DTRI_MAXLEVEL - i;
+    offset_index = end_level - i;
+    /* Get the local index of T's ancestor on level i */
+    local_index = (id >> (T8_DTRI_DIM * offset_index)) & children_m1;
+    /* Get the type and cube-id of T's ancestor on level i */
+    cid = t8_dtri_parenttype_Iloc_to_cid[type][local_index];
+    type = t8_dtri_parenttype_Iloc_to_type[type][local_index];
+    t->x |= (cid & 1) ? 1 << offset_coords : 0;
+    t->y |= (cid & 2) ? 1 << offset_coords : 0;
+#ifdef T8_DTRI_TO_DTET
+    t->z |= (cid & 4) ? 1 << offset_coords : 0;
+#endif
```

Comment:
The idea of this function is to avoid a call to t8_dtri_init_linear_id, as we have already computed the coordinates of t up to the level where t switches the shape from a pyramid to a tetrahedron. From this step on we can compute it as tet.
This is also the reason why we cannot use the original t8_dtri_init_linear_id, because the computation is different. 

To avoid further confusion, especially for a user who only looks at the tet/tri files I suggest to move this function to the pyramid-files. Would you agree with that?

Reply:
Ok, adding the assertions makes this function fail. I will work on a proper check before merging. We should move it to the pyramid-files anyway in my opinion

Reply:
Is fixed. 
After personal talk with @holke we decided to leave it in the tri-file, as it is works on tets/tris. However we added the remark in the documentation that it is mainly used by pyramids.